File: BabyTracker/Services/Auth/AuthService.swift
import Foundation
import SwiftData
import AuthenticationServices
import CryptoKit

/// Manages authentication state and the current caregiver identity.
/// Supports:
///   - Sign in with Apple (primary)
///   - Anonymous / guest mode (no account required to start using the app)
///   - Account linking (upgrade anonymous → Apple ID without losing data)

@MainActor
@Observable
final class AuthService {

    enum AuthState {
        case loading
        case anonymous(Caregiver)
        case authenticated(Caregiver)
    }

    private(set) var authState: AuthState = .loading
    private let modelContext: ModelContext
    private let supabase: SupabaseService

    var currentCaregiver: Caregiver? {
        switch authState {
        case .anonymous(let c), .authenticated(let c): return c
        case .loading: return nil
        }
    }

    var isAuthenticated: Bool {
        if case .authenticated = authState { return true }
        return false
    }

    init(modelContext: ModelContext, supabase: SupabaseService) {
        self.modelContext = modelContext
        self.supabase = supabase
    }

    // MARK: - Startup

    func resolveInitialAuthState() async {
        // 1. Check if there's a local caregiver marked as current device
        if let existing = try? fetchCurrentDeviceCaregiver() {
            authState = supabase.isAuthenticated ? .authenticated(existing) : .anonymous(existing)
            return
        }
        // 2. First launch — create anonymous caregiver, prompt for name
        let anonymous = Caregiver(displayName: "Parent", role: .admin, isCurrentDevice: true)
        modelContext.insert(anonymous)
        try? modelContext.save()
        authState = .anonymous(anonymous)
    }

    // MARK: - Sign In with Apple

    func handleAppleSignIn(result: Result<ASAuthorization, Error>) async {
        switch result {
        case .failure(let error):
            print("[Auth] Apple Sign In failed: \(error.localizedDescription)")
        case .success(let authorization):
            guard let credential = authorization.credential as? ASAuthorizationAppleIDCredential,
                  let tokenData = credential.identityToken,
                  let idToken = String(data: tokenData, encoding: .utf8),
                  let nonce = currentNonce else { return }

            do {
                try await supabase.signInWithApple(idToken: idToken, nonce: nonce)

                // Update the local caregiver with the Supabase user ID
                if let caregiver = currentCaregiver, let userID = supabase.currentUserID {
                    caregiver.supabaseUserID = userID
                    caregiver.syncStatus = .pending
                    if let fullName = credential.fullName {
                        let name = [fullName.givenName, fullName.familyName]
                            .compactMap { $0 }
                            .joined(separator: " ")
                        if !name.isEmpty { caregiver.displayName = name }
                    }
                    try? modelContext.save()
                    authState = .authenticated(caregiver)
                }
            } catch {
                print("[Auth] Supabase Apple sign in failed: \(error)")
            }
        }
    }

    // MARK: - Anonymous → Authenticated upgrade

    /// Upgrade an anonymous account to a full account without losing local data.
    /// All locally-created babies and entries retain their IDs — the sync manager
    /// will push them to Supabase after auth is established.
    func linkAppleAccount(result: Result<ASAuthorization, Error>) async {
        // Same flow as sign-in; local data is preserved because we keep the same
        // local Caregiver object — only the supabaseUserID field is updated.
        await handleAppleSignIn(result: result)
    }

    func signOut() async throws {
        try await supabase.signOut()

        // Keep local data; just clear the auth link
        if let caregiver = currentCaregiver {
            caregiver.supabaseUserID = nil
            caregiver.syncStatus = .pending
            try? modelContext.save()
            authState = .anonymous(caregiver)
        }
    }

    // MARK: - Nonce (for Sign In with Apple)

    private(set) var currentNonce: String?

    func prepareNonce() -> String {
        let nonce = randomNonceString()
        currentNonce = nonce
        return sha256(nonce)
    }

    private func randomNonceString(length: Int = 32) -> String {
        let charset = Array("0123456789ABCDEFGHIJKLMNOPQRSTUVXYZabcdefghijklmnopqrstuvwxyz-._")
        return String((0..<length).map { _ in charset.randomElement()! })
    }

    private func sha256(_ input: String) -> String {
        let data = Data(input.utf8)
        let hash = SHA256.hash(data: data)
        return hash.compactMap { String(format: "%02x", $0) }.joined()
    }

    // MARK: - Helpers

    private func fetchCurrentDeviceCaregiver() throws -> Caregiver? {
        let descriptor = FetchDescriptor<Caregiver>(
            predicate: #Predicate { $0.isCurrentDevice }
        )
        return try modelContext.fetch(descriptor).first
    }
}

File: flutter_app/lib/services/auth_service.dart
import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart';
import 'package:drift/drift.dart' show Value;
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../database/app_database.dart';
import '../models/caregiver.dart';
import '../models/sync_types.dart';
import 'supabase_service.dart';

enum AuthState { loading, anonymous, authenticated }

class AuthNotifier extends StateNotifier<AuthState> {
  final AppDatabase _db;
  final SupabaseService _supabase;

  /// The caregiver for the current device. Null only before [resolveInitialState]
  /// completes or after [signOut].
  Caregiver? currentCaregiver;

  AuthNotifier(this._db, this._supabase) : super(AuthState.loading);

  /// Called once at app startup (e.g. in main.dart after ProviderScope is live).
  Future<void> resolveInitialState() async {
    final dbCaregiver = await _db.getCurrentCaregiver();
    if (dbCaregiver != null) {
      currentCaregiver = Caregiver(
        id: dbCaregiver.id,
        displayName: dbCaregiver.displayName,
        email: dbCaregiver.email,
        supabaseUserId: dbCaregiver.supabaseUserId,
        role: CaregiverRole.values.byName(dbCaregiver.role),
        isCurrentDevice: true,
        syncStatus: SyncStatus.values.byName(dbCaregiver.syncStatus),
        remoteId: dbCaregiver.remoteId,
        createdAt: dbCaregiver.createdAt,
        updatedAt: dbCaregiver.updatedAt,
      );
      state = dbCaregiver.supabaseUserId != null
          ? AuthState.authenticated
          : AuthState.anonymous;
    } else {
      state = AuthState.anonymous;
    }
  }

  /// Sign in with Apple (iOS only). The raw [identityToken] comes from
  /// the sign_in_with_apple package.
  Future<void> signInWithApple(String identityToken) async {
    final nonce = _generateNonce();
    final hashedNonce = _sha256(nonce);
    final userId = await _supabase.signInWithApple(
      identityToken: identityToken,
      nonce: hashedNonce,
    );
    if (userId != null && currentCaregiver != null) {
      currentCaregiver = Caregiver(
        id: currentCaregiver!.id,
        displayName: currentCaregiver!.displayName,
        email: currentCaregiver!.email,
        supabaseUserId: userId,
        role: currentCaregiver!.role,
        isCurrentDevice: true,
        syncStatus: SyncStatus.pending,
        remoteId: currentCaregiver!.remoteId,
        createdAt: currentCaregiver!.createdAt,
        updatedAt: DateTime.now(),
      );
      // Persist the updated supabaseUserId to the local DB.
      await _db.upsertCaregiver(
        CaregiversCompanion(
          id: Value(currentCaregiver!.id),
          displayName: Value(currentCaregiver!.displayName),
          email: Value(currentCaregiver!.email),
          supabaseUserId: Value(userId),
          role: Value(currentCaregiver!.role.name),
          isCurrentDevice: const Value(true),
          syncStatus: const Value('pending'),
          remoteId: Value(currentCaregiver!.remoteId),
          createdAt: Value(currentCaregiver!.createdAt),
          updatedAt: Value(currentCaregiver!.updatedAt),
        ),
      );
      state = AuthState.authenticated;
    }
  }

  Future<void> signOut() async {
    await _supabase.signOut();
    currentCaregiver = null;
    state = AuthState.anonymous;
  }

  bool get isAuthenticated => state == AuthState.authenticated;
  bool get isLoading => state == AuthState.loading;

  // ── Helpers ───────────────────────────────────────────────────────────────

  String _generateNonce([int length = 32]) {
    const chars =
        '0123456789ABCDEFGHIJKLMNOPQRSTUVXYZabcdefghijklmnopqrstuvwxyz-._';
    final random = Random.secure();
    return List.generate(
      length,
      (_) => chars[random.nextInt(chars.length)],
    ).join();
  }

  String _sha256(String input) {
    final bytes = utf8.encode(input);
    final digest = sha256.convert(bytes);
    return digest.toString();
  }
}
